Cannot package project with non-importable dependencies

## :bug: Bug Report

### Affected Languages
<!--
Check the box (with an X) for any language runtime that you know is affected by
the reported bug. If you're uncertain whether a language is affected or not,
please leave the bux un-checked.
-->
- [x] `TypeScript` or `Javascript`
- [x] `Python`
- [x] `Java`
- [x] .NET (`C#`, `F#`, ...)
- [x] `Go`

### General Information
* **JSII Version:** jsii 1.49.0
* **Platform:** Unix

### What is the problem?
<!--
Describe in as much detail as possible the problem that you have identified.
Whenever possible, add reference to the minimal reproduction of the issue, as
this will greatly help with identifying the root cause of the problem, and
ensuring a candidate fix is working.

If your minimal reprodiction is large, you may provide references to a [gist][1]
or any other publicly viewable location.
-->

If a JSII project takes a dependency on a package that is not importable, `jsii-pacmak` fails. A package may not be importable if it only provides command-line scripts, for example the [shx](https://www.npmjs.com/package/shx) package. shx does not specify the "main" field in package.json, and it does not have a default "index.js" file that serves as an entrypoint -- its main utility is through the "bin" field which directly runs JS code.

This previously worked, but stopped working when we upgraded our project from JSII v1.47.0 to v1.49.0.

Our use case for taking "non-importable" libraries as dependencies is to bundle them along with our project, so that the command-line scripts in "shx" can be used for functionality in our construct library.

Related issue: https://github.com/projen/projen/pull/1336

### Verbose Log
<!--
* If your issue is with `jsii`: please provide the output of running `jsii` with
  the `--verbose` option.
* If your issue is with `jsii-pacmak`: please provide the output of running
  `jsii-pacmak` with the `--verbose` option.
* If your issue is when running generated code: please provide the output of
  running your program with the `JSII_DEBUG=1` environment variable being set.

If the output is very large, you may store it in a [gist][1]. In any case,
please review the contents of the log you are about to submit and edit any
confidential information out.

[1]: https://gist.github.com
-->

Reproduction repo: https://github.com/Chriscbr/jsii-pacmak-repro

```
$ npx jsii-pacmak
[jsii-pacmak] [WARN] Exception occurred, not cleaning up
[jsii-pacmak] [WARN] java failed
Error: Cannot find module 'shx'
Require stack:
- /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/util.js
- /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/packaging.js
- /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/npm-modules.js
- /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/index.js
- /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/bin/jsii-pacmak.js
- /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/bin/jsii-pacmak
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.resolve (node:internal/modules/cjs/helpers:108:19)
    at Object.findDependencyDirectory (/Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/util.js:18:32)
    at /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/dependency-graph.js:46:42
    at Array.map (<anonymous>)
    at real$traverseDependencyGraph (/Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/dependency-graph.js:45:10)
    at async Object.findLocalBuildDirs (/Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/target.js:76:5)
    at async /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/targets/java.js:171:29
    at async JavaBuilder.generateMavenSettingsForLocalDeps (/Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/targets/java.js:173:51)
    at async /Volumes/workplace/jsii-pacmak-repro/node_modules/jsii-pacmak/lib/targets/java.js:103:1
 ```

--- a/packages/jsii-pacmak/lib/dependency-graph.ts
+++ b/packages/jsii-pacmak/lib/dependency-graph.ts
@@ -58,6 +58,7 @@ export interface TraverseDependencyGraphHost {
 export interface PackageJson {
   readonly dependencies?: { readonly [name: string]: string };
   readonly peerDependencies?: { readonly [name: string]: string };
+  readonly bundleDependencies?: string[];
   readonly bundledDependencies?: string[];
 
   readonly [key: string]: unknown;
@@ -93,7 +94,7 @@ async function real$traverseDependencyGraph(
       .filter(
         (m) =>
           !util.isBuiltinModule(m) &&
-          !(meta.bundledDependencies ?? []).includes(m),
+          (meta.bundledDependencies ?? meta.bundleDependencies)?.includes(m),
       )
       .map(async (dep) => {
         const dependencyDir = await host.findDependencyDirectory(

--- a/packages/jsii-pacmak/lib/dependency-graph.ts
+++ b/packages/jsii-pacmak/lib/dependency-graph.ts
@@ -94,7 +94,7 @@ async function real$traverseDependencyGraph(
       .filter(
         (m) =>
           !util.isBuiltinModule(m) &&
-          (meta.bundledDependencies ?? meta.bundleDependencies)?.includes(m),
+          !(meta.bundledDependencies ?? meta.bundleDependencies)?.includes(m),
       )
       .map(async (dep) => {
         const dependencyDir = await host.findDependencyDirectory(

--- a/packages/jsii-pacmak/lib/dependency-graph.ts
+++ b/packages/jsii-pacmak/lib/dependency-graph.ts
@@ -58,6 +58,7 @@ export interface TraverseDependencyGraphHost {
 export interface PackageJson {
   readonly dependencies?: { readonly [name: string]: string };
   readonly peerDependencies?: { readonly [name: string]: string };
+  readonly bundledDependencies?: string[];
 
   readonly [key: string]: unknown;
 }
@@ -88,7 +89,12 @@ async function real$traverseDependencyGraph(
   ]);
   return Promise.all(
     Array.from(deps)
-      .filter((m) => !util.isBuiltinModule(m))
+      // No need to pacmak the dependency if it's built-in, or if it's bundled
+      .filter(
+        (m) =>
+          !util.isBuiltinModule(m) &&
+          !(meta.bundledDependencies ?? []).includes(m),
+      )
       .map(async (dep) => {
         const dependencyDir = await host.findDependencyDirectory(
           dep,

--- a/packages/jsii-pacmak/lib/dependency-graph.ts
+++ b/packages/jsii-pacmak/lib/dependency-graph.ts
@@ -94,7 +94,8 @@ async function real$traverseDependencyGraph(
       .filter(
         (m) =>
           !util.isBuiltinModule(m) &&
-          !(meta.bundledDependencies ?? meta.bundleDependencies)?.includes(m),
+          !meta.bundledDependencies?.includes(m) &&
+          !meta.bundleDependencies?.includes(m),
       )
       .map(async (dep) => {
         const dependencyDir = await host.findDependencyDirectory(

--- a/packages/jsii-pacmak/test/dependency-graph.test.ts
+++ b/packages/jsii-pacmak/test/dependency-graph.test.ts
@@ -113,3 +113,41 @@ test('stops traversing when callback returns false', async () => {
   expect(mockHost.readJson).toHaveBeenCalledTimes(2);
   expect(mockHost.findDependencyDirectory).toHaveBeenCalledTimes(1);
 });
+
+test('dont call findDependencyDirectory for bundledDependencies', async () => {
+  // GIVEN the following package dependency graph:
+  //    A -> B -> C
+  const packages: Record<string, { root: string; meta: any }> = {
+    A: {
+      root: join(tmpdir(), 'A'),
+      meta: { dependencies: { B: '*' }, bundledDependencies: ['B'] },
+    },
+  };
+
+  const cb: Callback = jest.fn().mockName('callback').mockReturnValue(true);
+
+  mockHost.readJson.mockImplementation((file) => {
+    const result = Object.values(packages).find(
+      ({ root }) => file === join(root, 'package.json'),
+    )?.meta;
+    return result != null
+      ? Promise.resolve(result)
+      : Promise.reject(new Error(`Unexpected file access: ${file}`));
+  });
+
+  mockHost.findDependencyDirectory.mockImplementation(async (dep, _dir) => {
+    const result = packages[dep]?.root;
+    if (result == null) {
+      throw new Error(`Unknown dependency: ${dep}`);
+    }
+    return result;
+  });
+
+  // WHEN
+  await expect(
+    traverseDependencyGraph(packages.A.root, cb, mockHost),
+  ).resolves.not.toThrow();
+
+  // THEN
+  expect(mockHost.findDependencyDirectory).not.toHaveBeenCalled();
+});

--- a/packages/jsii-pacmak/test/dependency-graph.test.ts
+++ b/packages/jsii-pacmak/test/dependency-graph.test.ts
@@ -115,8 +115,6 @@ test('stops traversing when callback returns false', async () => {
 });
 
 test('dont call findDependencyDirectory for bundledDependencies', async () => {
-  // GIVEN the following package dependency graph:
-  //    A -> B -> C
   const packages: Record<string, { root: string; meta: any }> = {
     A: {
       root: join(tmpdir(), 'A'),
@@ -126,22 +124,7 @@ test('dont call findDependencyDirectory for bundledDependencies', async () => {
 
   const cb: Callback = jest.fn().mockName('callback').mockReturnValue(true);
 
-  mockHost.readJson.mockImplementation((file) => {
-    const result = Object.values(packages).find(
-      ({ root }) => file === join(root, 'package.json'),
-    )?.meta;
-    return result != null
-      ? Promise.resolve(result)
-      : Promise.reject(new Error(`Unexpected file access: ${file}`));
-  });
-
-  mockHost.findDependencyDirectory.mockImplementation(async (dep, _dir) => {
-    const result = packages[dep]?.root;
-    if (result == null) {
-      throw new Error(`Unknown dependency: ${dep}`);
-    }
-    return result;
-  });
+  fakeReadJson(packages);
 
   // WHEN
   await expect(
@@ -151,3 +134,70 @@ test('dont call findDependencyDirectory for bundledDependencies', async () => {
   // THEN
   expect(mockHost.findDependencyDirectory).not.toHaveBeenCalled();
 });
+
+test('dont call findDependencyDirectory for bundleDependencies', async () => {
+  const packages: Record<string, { root: string; meta: any }> = {
+    A: {
+      root: join(tmpdir(), 'A'),
+      meta: { dependencies: { B: '*' }, bundleDependencies: ['B'] },
+    },
+  };
+
+  const cb: Callback = jest.fn().mockName('callback').mockReturnValue(true);
+
+  fakeReadJson(packages);
+
+  // WHEN
+  await expect(
+    traverseDependencyGraph(packages.A.root, cb, mockHost),
+  ).resolves.not.toThrow();
+
+  // THEN
+  expect(mockHost.findDependencyDirectory).not.toHaveBeenCalled();
+});
+
+test('dont call findDependencyDirectory for bundleDependencies AND bundledDependencies', async () => {
+  const packages: Record<string, { root: string; meta: any }> = {
+    A: {
+      root: join(tmpdir(), 'A'),
+      meta: {
+        dependencies: { B: '*', C: '*' },
+        bundleDependencies: ['B'],
+        bundledDependencies: ['C'],
+      },
+    },
+  };
+
+  const cb: Callback = jest.fn().mockName('callback').mockReturnValue(true);
+
+  fakeReadJson(packages);
+
+  // WHEN
+  await expect(
+    traverseDependencyGraph(packages.A.root, cb, mockHost),
+  ).resolves.not.toThrow();
+
+  // THEN
+  expect(mockHost.findDependencyDirectory).not.toHaveBeenCalled();
+});
+
+function fakeReadJson(
+  fakePackages: Record<string, { root: string; meta: any }>,
+) {
+  mockHost.readJson.mockImplementation((file) => {
+    const result = Object.values(fakePackages).find(
+      ({ root }) => file === join(root, 'package.json'),
+    )?.meta;
+    return result != null
+      ? Promise.resolve(result)
+      : Promise.reject(new Error(`Unexpected file access: ${file}`));
+  });
+
+  mockHost.findDependencyDirectory.mockImplementation(async (dep, _dir) => {
+    const result = fakePackages[dep]?.root;
+    if (result == null) {
+      throw new Error(`Unknown dependency: ${dep}`);
+    }
+    return result;
+  });
+}

